File: Problems/Hackerrank/leftRotation.js
/****************************************************************/
/*Problem: Left rotation (HR)  ********/
/****************************************************************/
/*
-- Summary:
A left rotation operation on an array of size n shifts each of the array's elements 1 unit to the left. For example, if 2 left rotations are performed on array [1,2,3,4,5], then the array would become [3,4,5,1,2]. Given an array of n integers and a number, d, perform d left rotations on the array. Then print the updated array as a single line of space-separated integers.
-- Input(s):
First line contains two space-separated integers denoting the respective values of n (the number of integers) and d (the number of left rotations you must perform). The second line contains n space-separated integers describing the respective elements of the array's initial state.
-- Expected output(s):
Print a single line of n space-separated integers denoting the final state of the array after performing d left rotations.
-- Constraints:
1<=n<=10^5 | 1<=d<=10^5 | 1<=ai<=10^5
*/
function leftRotation (n,d,a) {
  let endArray = [];
  let modArray = [];

  //Gets elements to be shifted:
  endArray = a.slice(0,d); 
  //Removes elements to be shifted from original array and concatenates them at the end:
  endArray = a.splice(d).concat(endArray); 
  //Converts array to string:
  endArray = endArray.toString();
  //Replaces the commas with spaces:
  endArray = endArray.split(',').join(' ');

  return endArray;
};
console.log(leftRotation(5,2,[1,2,3,4,5]));
console.log(leftRotation(5,4,[1,2,3,4,5]));
console.log(leftRotation(5,1,[1,2,3,4,5]));

File: Problems/Hackerrank/leftRotation.cpp
/*********************************************************************************************/
/* Problem: Left Rotation (HR) ********/
/*********************************************************************************************/
/*
--Problem statement:
  A left rotation operation on an array of size n shifts each of the array's elements 1 unit to
  the left. Given an integer, d, rotate the array that many steps left and return the result.

  > Example:
      d = 2
      arr = [1,2,3,4,5]
      After 2 rotations, arr' = [3,4,5,1,2]

  > Input:
      int d: the amount to rotate by
      int arr[n]: the array to rotate

  > Output - int[n]: the rotated array

  > Constraints:
      1<=n<=10^5
      1<=d<=n
      1<=a[i]<=10^6

--Reasoning:
  Perform 'd' iterations on the array adding all elements which would be rotated to the back of the
  array. After that, remove these 'd' elements from their original positions. What will be left is
  the rotated array.

--Time complexity: O(d), where 'd' is the number of rotations.

--Space complexity: O(1), there is no dependency on the input size.
*/

#include <iostream>
#include <vector>

void displayArray(const std::vector<int> &arr);

void rotateLeft(int d, std::vector<int> &arr)
{
  for (int i{0}; i < d; i++)
  {
    arr.push_back(arr.at(i));
  }

  arr.erase(arr.begin(), arr.begin() + d);
}

int main()
{
  std::vector<int> vec{1, 2, 3, 4, 5};
  int d = 4;
  rotateLeft(d, vec);
  displayArray(vec); // Expected: 5 1 2 3 4

  vec = {41, 73, 89, 7, 10, 1, 59, 58, 84, 77, 77, 97, 58, 1, 86, 58, 26, 10, 86, 51};
  d = 10;
  rotateLeft(d, vec);
  displayArray(vec); // Expected: 77 97 58 1 86 58 26 10 86 51 41 73 89 7 10 1 59 58 84 77

  return 0;
}

void displayArray(const std::vector<int> &arr)
{
  for (const auto &a : arr)
  {
    std::cout << a << "  ";
  }
  std::cout << "\n";
}